signer/xpi/x509.go

```diff
@@ -104,3 +133,99 @@ func (s *PKCS7Signer) MakeEndEntity(cn string) (eeCert *x509.Certificate, eeKey
 	}
 	return
 }
+
+
+// KeyType is the type to use in keyOptions to tell MakeDEREndEntity
+// which type of crypto.PrivateKey to generate
+type KeyType int
+const (
+	// keyTypeRSA is the keyOptions.keyType to generate an rsa.PrivateKey
+	keyTypeRSA KeyType = iota
+
+	// keyTypeECDSA is the keyOptions.keyType to generate an ecdsa.PrivateKey
+	keyTypeECDSA KeyType = iota
+)
+
+
+func getKeyOptionsForCOSEAlg(alg *cose.Algorithm) (*keyOptions) {
+	switch alg.Name {
+	case "ES256":
+		return &keyOptions{
+			keyType: keyTypeECDSA,
+			ecdsaCurve: elliptic.P256(),
+		}
+	case "ES384":
+		return &keyOptions{
+			keyType: keyTypeECDSA,
+			ecdsaCurve: elliptic.P384(),
+		}
+	case "ES512":
+		return &keyOptions{
+			keyType: keyTypeECDSA,
+			ecdsaCurve: elliptic.P521(),
+		}
+	case "PS256":
+		return &keyOptions{
+			keyType: keyTypeRSA,
+			rsaBits: 2048, // TODO: fixme
+		}
+	default:
+		return nil
+	}
+}
+
+// keyOptions
+type keyOptions struct {
+	keyType KeyType
+
+	// rsaBits is the bit length or size of the RSA key to generate only applies when keyType is keyTypeRSA
+	rsaBits int
+
+	// ecdsaCurve is the elliptic curve to generate an ECDSA key with only applies when keyType is keyTypeECDSA
+	ecdsaCurve elliptic.Curve
+}
+
+// MakeDEREndEntity generates an EE cert and private key like
+// MakeEndEntity but returns the DER encoded certificate bytes instead
+// of a X.509 certificate and takes additional options to generate EEs
+// using different key types or params than the issuer cert
```

Comment:
the only difference between a DER and an X.509 is the call to `eeCert, err = x509.ParseCertificate(derCert)`, so.... dedup?